File: Algorithms/0954_Array_of_Doubled_Pairs/Python/Array_of_Doubled_Pairs_Solution_1.py

# Space: O(n)
# Time: O(n^2)
import collections


class Solution:
    def canReorderDoubled(self, arr):
        length = len(arr)
        if length == 2:
            return True if arr[0] * 2 == arr[1] or arr[1] * 2 == arr[0] else False
        cache = collections.defaultdict(list)
        res = []
        for i in range(length):
            cache[arr[i]].append(i)

        less_than_zero = [n for n in arr if n < 0]
        greater_than_zero = [n for n in arr if n >= 0]
        less_than_zero = sorted(less_than_zero, reverse=True)
        greater_than_zero = sorted(greater_than_zero)
        arr_copy = less_than_zero + greater_than_zero

        for n in arr_copy:
            target_number = 2 * n
            if n in cache and target_number in cache:
                if cache[n]:
                    res.append(n)
                    cache[n].pop()
                    if not cache[n]: cache.pop(n)
                if cache[target_number]:
                    res.append(target_number)
                    cache[target_number].pop()
                    if not cache[target_number]: cache.pop(target_number)

        return sorted(res) == sorted(arr)

File: Algorithms/0954_Array_of_Doubled_Pairs/Java/ArrayofDoubledPairsSolution1.java

// Space: O(n)
// Time: O(n^2)

class Solution {
    private Map<Integer, Integer> cache = null;
    public boolean canReorderDoubled(int[] arr) {
            if (arr.length == 2) {
            if (arr[0] * 2 == arr[1] || arr[1] * 2 == arr[0]) return true;
            return false;
        }
        cache = new HashMap<>();
        for (int i : arr) {
            cache.put(i, cache.getOrDefault(i, 0) + 1);
        }

        Integer[] temp = new Integer[arr.length];
        for (int i = 0; i < arr.length; ++i)
            temp[i] = arr[i];
        Arrays.sort(temp, Comparator.comparingInt(Math::abs));


        for (Integer i : temp) {
            if (cache.get(i) == 0) continue;
            if (cache.getOrDefault(2 * i, 0) == 0) return false;
            cache.put(i, cache.get(i) - 1);
            cache.put(i * 2, cache.get(i * 2) - 1);
        }
        return true;
    }
}




